### ISSUE TITLE
Cropping/panscan positioning/clipping are completely broken (with and without use_margins)

### ISSUE THREAD

**Issue by wm4** [libass/libass#218]
I can only give a big problem statement. I've seen issues with this since forever.

This affects usage of ass_set_margins() and ass_set_use_margins(). Here are some things that go wrong:
- if use_margins is set, normal text should be completely unaffected by margins, but testing shows that it gets moved and clipped in nonsensical ways
- even with normal panscan usage, an event emulating both subtitles and toptitles at the same time (e.g. `toptitle\N\N...\N\Nsubtitle`), the top can be clipped off in weird ways, when it should use the entire screen space
- there are probably additional problems with anamorphic video
- even if use_margins is disabled, margins affect normal subtitles (this could be a feature, not a bug - e.g. if panscan crops of the sides of the video, the subtitles should be broken to be within the cropped video rectangle)
- apparently there are no issues with explicitly positioned events

"Someone" should investigate and fix these.

mpv in particular can make elaborate use of it, including cropping uneven amounts of the video frame. One can use this set of mpv key bindings for testing:

```
b add video-zoom 0.1
B add video-zoom -0.1
ctrl+left add video-pan-x 0.1
ctrl+right add video-pan-x -0.1
ctrl+up add video-pan-y 0.1
ctrl+down add video-pan-y -0.1
```

**Comment by astiob** [comment]
I did some testing with (non-anamorphic) ASS only, with and without `--ass-force-margins`.

**Firstly,** just confirming your findings:

For explicitly positioned events, whatever panning and scanning I do, the subtitles act as if they were burned onto the video. (In particular, if I pan so that part of the video is cropped off, the subtitles are cropped off as well.) This seems desirable. (The only problem is that when zooming in a lot, libass starts refusing to rasterize some outlines because they become too large.)

`--ass-force-margins` seems to affect only the placement/anchoring of normal subtitles, i. e. whether they’re positioned relative to the video or to the window; but they’re clipped and moved regardless of this option.

**Now,** after some testing and thinking, I think I’ve figured out how the seemingly weird movement and clipping works. The basic idea is that libass tries to fit the text into the visible part of the video (the intersection of the video rectangle and the window rectangle). To this end, it performs line alignment and line wrapping based on the dimensions of the visible part of the video. As the visible width shrinks, text is wrapped into progressively shorter lines, eventually consisting of just one word each and getting clipped when those words don’t fit into the visible width. Centred text (both horizontally and vertically) is centred within the visible part of the video, not the entire video or the entire window.

The most curious/nonsensical part is that ASS margins—MarginL, MarginR and MarginV, which define where noncentred lines are positioned—get _scaled_ by libass from the full video size to the size of the visible portion. Hence, the more of the video is off screen, the smaller the margins and the closer to the edge of the window text is placed (even if the off-screen part is on the opposite side of the window [which can only happen when margins are used]). Text also appears to be clipped to the size of the currently visible portion of the video (again, even if it is on the opposite side of the window).

I think this explains all the weirdness that I’ve seen. I hope I’m not forgetting anything.

----

What do we _want_ to happen to normal text during panscan? Do we want to even move the text at all, or should it just stay anchored to the window? Perhaps re-layout the text when zooming but not when panning? But I’m not sure how libass could distinguish between the two. I can also see how it makes sense for normal text subtitles to be completely unaffected by panscan whatsoever. When margin use is enabled, that is. I’m not sure what’s the right course of action with disabled margins. Perhaps handle all subtitles like explicitly positioned ones?

In any case, I think the fact that ASS margins are scaled from the full video size to the visible part doesn’t make much sense. It does allow us to fit lines into the visible video when the visible video size is very small, but at the cost of being nonintuitive the rest of the time. I would just keep the margins alone when panning (but still scale them by the zoom factor).

The fact that text is clipped to the visible video size even when it’s on the opposite side of the screen should also viewed as a bug and fixed.

**Comment by astiob** [comment]
While we’re on the topic of margins, I think we should also make use of the margins to the left and right of the video, e. g. when playing a 4:3 video on a wide screen.

**Comment by wm4** [comment]
Thanks for your analysis.

I think the following behavior would make sense:
- If margins are not enabled, then strictly put subs over the video, even if it means that text gets cropped off, or that non-video space isn't used.
- If margins are enabled, use the whole screen for text and ignore the video rectangle.

That means we would remove the code that tries to fit text into the screen/video intersection, which should be much simpler and more predictable. This is what I meant by dropping obscure features.

Also in case margins are enabled, whether to fix an event to the video or whether to use the full screen should not just depend on `\pos`, but on other things as well. Possibly reusing the style override heuristic.

**Comment by astiob** [comment]
Yep, sounds good to me.

**Comment by astiob** [comment]
Fixed by #385.
